### PR TITLE
Add `make deb` support for aarch64

### LINKED ISSUE #27045: Feature Request: aarch64 `make deb` support
As the title suggests, would be great to be able to build deb packages ('make deb' target) for aarch64 (for example the [odroid c2](http://www.hardkernel.com/main/products/prdt_info.php?g_code=G145457216438&tab_idx=2)

I have made an initial attempt based on the existing armhf support, but in this case I don't really know what I'm doing (the `make deb` build still hangs), and with the lack of aarch64 support in other areas (go for example?) the level of effort might be more significant than I believe at present.

### PR DESCRIPTION
**Pull request by hqhq**
**\- What I did**
Add `make deb` support for aarch64.

**\- How I did it**
Added Dockerfiles which can be used to build deb packages.

**\- How to verify it**
run `make deb` on arm64 service and check deb packages it generated.

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

Fixes: #27045

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

**Comment by hqhq** [comment]
@tophj-ibm @vielmetti Can you help test it out, I've tested trusty but not able to test xenial out, I always got bunch of errors like:
```
E: Failed to fetch http://ports.ubuntu.com/ubuntu-ports/pool/main/p/python3.5/libpython3.5-stdlib_3.5.2-2~16.04_arm64.deb  Writing more data than expected (2014047 > 2013750)

E: Failed to fetch http://ports.ubuntu.com/ubuntu-ports/pool/main/p/perl/perl-modules-5.22_5.22.1-9_all.deb  Writing more data than expected (2815858 > 2814692)

E: Failed to fetch http://ports.ubuntu.com/ubuntu-ports/pool/main/p/perl/libperl5.22_5.22.1-9_arm64.deb  Writing more data than expected (2814989 > 2814692)

E: Failed to fetch http://ports.ubuntu.com/ubuntu-ports/pool/main/i/icu/libicu55_55.1-7_arm64.deb  Writing more data than expected (7485051 > 7484754)

E: Failed to fetch http://ports.ubuntu.com/ubuntu-ports/pool/main/l/linux/linux-libc-dev_4.4.0-45.66_arm64.deb  Writing more data than expected (5443601 > 5442230)
```
When build the Dockerfile, I googled and lots of people met the same issue on xenial but I didn't find a solution, not sure if it's related to my network(I'm behind proxy).

File: contrib/builder/deb/aarch64/generate.sh [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+			packages+=( libsystemd-journal-dev )
+			# aarch64 doesn't have an official downloadable binary for go.
+			# And gccgo for trusty only includes Go 1.2 implementation which
+			# is too old to build current go source, fortunately trusty has
+			# golang-1.6-go package can be used as bootstrap.
+			packages+=( golang-1.6-go )
+			;;
+		xenial)
+			packages+=( libsystemd-dev )
+			packages+=( gccgo libseccomp-dev)
```
Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/generate.sh]: nit: gccgo should be golang-go http://packages.ubuntu.com/xenial/golang-go. While this should work for now, gccgo doesn't seem to be updated very much, for instance being go 1.2 on trusty.

Comment (hqhq) [inline contrib/builder/deb/aarch64/generate.sh]: gccgo for xenial should be gccgo5 and that can work as bootstrap, but sure `golang-go` seems to be a better choice, I'll update PR, thanks.

**Comment by hqhq** [comment]
Updated.

File: contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Install Go
+# aarch64 doesn't have official go binaries, so use the version of go installed from
+# the image to build go from source.
+ENV GO_VERSION 1.7.3
+RUN mkdir /usr/src/go && curl -fsSL https://storage.googleapis.com/golang/go${GO_VERSION}.src.tar.gz | tar -v -C /usr/src/go -xz --strip-components=1 \
+	&& cd /usr/src/go/src \
+	&& GOOS=linux GOARCH=arm64 GOROOT_BOOTSTRAP="$(go env GOROOT)" ./make.bash
+
+ENV PATH /usr/src/go/bin:$PATH
+
```
Comment (StefanScherer) [inline contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile]: @hqhq I couldn't build it [and compared](https://github.com/docker/docker/blob/c5ceb0f945d390500a0045b14ffe4dcecfe46a85/contrib/builder/deb/armhf/ubuntu-trusty/Dockerfile#L9) the `contrib/builder/deb/aarch64/*/Dockerfile` with ` contrib/builder/deb/amrhf/*/Dockerfile`.
After adding

```Dockerfile
ENV AUTO_GOPATH 1
```

to all `contrib/builder/deb/aarch64/*/Dockerfile` I'm able to build the `docker-engine_*_arm64.deb` packages .

File: contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# Install Go
+# aarch64 doesn't have official go binaries, so use the version of go installed from
+# the image to build go from source.
+ENV GO_VERSION 1.7.3
+RUN mkdir /usr/src/go && curl -fsSL https://storage.googleapis.com/golang/go${GO_VERSION}.src.tar.gz | tar -v -C /usr/src/go -xz --strip-components=1 \
+	&& cd /usr/src/go/src \
+	&& GOOS=linux GOARCH=arm64 GOROOT_BOOTSTRAP="$(go env GOROOT)" ./make.bash
+
+ENV PATH /usr/src/go/bin:$PATH
+
```
Comment (StefanScherer) [inline contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile]: Add
```Dockerfile
ENV AUTO_GOPATH 1
```

File: contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# Install Go
+# aarch64 doesn't have official go binaries, so use the version of go installed from
+# the image to build go from source.
+ENV GO_VERSION 1.7.3
+RUN mkdir /usr/src/go && curl -fsSL https://storage.googleapis.com/golang/go${GO_VERSION}.src.tar.gz | tar -v -C /usr/src/go -xz --strip-components=1 \
+	&& cd /usr/src/go/src \
+	&& GOOS=linux GOARCH=arm64 GOROOT_BOOTSTRAP="$(go env GOROOT)" ./make.bash
+
+ENV PATH /usr/src/go/bin:$PATH
```
Comment (StefanScherer) [inline contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile]: should be reordered like the amd64 Dockerfile to reduce the diffs between the different architectures.

```
> ENV PATH $PATH:/usr/local/go/bin
20d13
< ENV PATH /usr/src/go/bin:$PATH
```

File: contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+ENV DOCKER_BUILDTAGS apparmor selinux pkcs11 seccomp
```
Comment (StefanScherer) [inline contrib/builder/deb/aarch64/ubuntu-xenial/Dockerfile]: should be reordered like the amd64 Dockerfile to reduce the diffs between the different architectures.

```
< ENV DOCKER_BUILDTAGS apparmor selinux pkcs11
---
> ENV DOCKER_BUILDTAGS apparmor pkcs11 selinux
```

**Review by StefanScherer — COMMENTED** [review]
Great approach. But I had to add to add

``` Dockerfile
ENV AUTO_GOPATH 1
```

and could build the deb packages.

**Comment by hqhq** [comment]
@StefanScherer Thanks for your test and feedback, I've updated per your comments, thanks.

**Comment by hqhq** [comment]
ping @justincormack @tianon I think it's good to review now, we have a tested-by :).

File: contrib/builder/deb/aarch64/generate.sh [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		*)
+			;;
+	esac
+
+	echo "# Install Go" >> "$version/Dockerfile"
+	echo "# aarch64 doesn't have official go binaries, so use the version of go installed from" >> "$version/Dockerfile"
+	echo "# the image to build go from source." >> "$version/Dockerfile"
+
+	awk '$1 == "ENV" && $2 == "GO_VERSION" { print; exit }' ../../../../Dockerfile.aarch64 >> "$version/Dockerfile"
+	echo 'RUN mkdir /usr/src/go && curl -fsSL https://storage.googleapis.com/golang/go${GO_VERSION}.src.tar.gz | tar -v -C /usr/src/go -xz --strip-components=1 \' >> "$version/Dockerfile"
```
Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/generate.sh]: small nit, can you change the link to https://golang.org/dl/go${GO_VERSION}.src.tar.gz to be more consistent with the rest of the Dockerfiles (I just went through and changed them all) 😄

File: contrib/builder/deb/aarch64/generate.sh [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			;;
+		xenial)
+			packages+=( libsystemd-dev )
+			packages+=( golang-go libseccomp-dev)
+
+			dockerBuildTags="$dockerBuildTags seccomp"
+			runcBuildTags="$runcBuildTags seccomp"
+			;;
+		*)
+			echo "Unsupported distro"
```
Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/generate.sh]: might be helpful to output $distro and $suite here too

**Comment by hqhq** [comment]
@tophj-ibm Done, thanks.

**Comment by szepeviktor** [comment]
Does it mean that apt.dockerproject.org will host the arm64.deb too?

File: contrib/builder/deb/aarch64/generate.sh [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		debhelper # for easy ".deb" building
+		dh-apparmor # for apparmor debhelper
+		dh-systemd # for systemd debhelper integration
+		git # for "git commit" info in "docker -v"
+		libapparmor-dev # for "sys/apparmor.h"
+		libdevmapper-dev # for "libdevmapper.h"
+		libltdl-dev # for pkcs11 "ltdl.h"
+		libsqlite3-dev # for "sqlite3.h"
+		pkg-config # for detecting things like libsystemd-journal dynamically
+	)
```
Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/generate.sh]: you're going to need to add in cmake and vim-common to the packages here, which are dependencies for tini. see #28212

**Comment by hqhq** [comment]
@tophj-ibm Done, thanks for tracking this.

@szepeviktor I don't know, let's hope so.

**Comment by hqhq** [comment]
ping @justincormack @tianon @thaJeztah

File: contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#
+# THIS FILE IS AUTOGENERATED; SEE "contrib/builder/deb/aarch64/generate.sh"!
+#
+
+FROM aarch64/ubuntu:trusty
+
+RUN apt-get update && apt-get install -y apparmor bash-completion btrfs-tools build-essential cmake curl ca-certificates debhelper dh-apparmor dh-systemd git libapparmor-dev libdevmapper-dev libltdl-dev libsqlite3-dev pkg-config vim-common libsystemd-journal-dev golang-1.6-go --no-install-recommends && rm -rf /var/lib/apt/lists/*
+
+RUN update-alternatives --install /usr/bin/go go /usr/lib/go-1.6/bin/go 100
```
Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile]: aarch64 doesn't have official go binaries, so they have to build go from source which requires an older version of go >1.4 to bootstrap, which is what is being installed here.

Comment (tophj-ibm) [inline contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile]: this version of go can probably be removed after building the one below though, in both dockerfiles.

**Comment by thaJeztah** [comment]
We (@andrewhsu, @mlaventure) were  just discussing this PR, and we're okay with merging this, but won't be publishing binaries for now, given that we don't have the hardware for this yet :smile:

File: contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile (hunk 9 above)
Comment (mlaventure) [inline contrib/builder/deb/aarch64/ubuntu-trusty/Dockerfile]: Oh, right, thanks for the clarification.

**Review by mlaventure — APPROVED** [review]
LGTM

**Review by tianon — APPROVED** [review]
:+1: :heart:

**Comment by thaJeztah** [comment]
Windows CI failure is not related here :smile:, merging :tada:

**Comment by luxas** [comment]
@thaJeztah Re hardware, I guess https://www.packet.net/ would be willing to support you with some boxes, but I don't know (I do not work there or anything)
Reach out to them and ask :)

I really would like to have automated aarch64 builds :+1:

**Comment by thaJeztah** [comment]
Thanks for the suggestion @luxas, @andrewhsu perhaps worth a try?

**Comment by zsmith928** [comment]
@thaJeztah yup we'll be happy to sponsor this work, just email me zac at packet dot net and I'll get it going.

**Comment by thaJeztah** [comment]
Sweet! Thanks @zsmith928 :heart:, I'll ask @andrewhsu to get into contact, as he's gonna be in charge of building packages :smile:

**Comment by andrewhsu** [comment]
@zsmith928 Yes! Let's get stuff going! I'll send you email.

**Pull request merged**